File: c/hash-table/bulls-and-cows.c
char* getHint(char* secret, char* guess) {
    int *ret;
    int a = 0;
    int b = 0;
    int len = strlen(guess);
    int m[256] = {0};

    ret = malloc(sizeof(int) * (len+2));

    for(int i = 0; i < len; ++i){
        if(secret[i] == guess[i]){
            ++a;
        }else{
            m[secret[i]]++;
        }
    }

    for(int i = 0; i < len; ++i){
        if(secret[i] != guess[i] && m[guess[i]]){
            m[guess[i]]--;
            ++b;
        }
    }

    len = sprintf(ret, "%dA%dB", a, b);
    ret[len] = 0;

    return ret;
}

File: python/hash-table/bulls-and-cows.py
class Solution(object):
    def getHint(self, secret, guess):
        d = {}
        bulls, cows = 0, 0
        for i in xrange(len(secret)):
            if secret[i] == guess[i]:
                bulls += 1
            else:
                if d.get(secret[i]):
                    d[secret[i]] += 1
                else:
                    d[secret[i]] = 1

        for i in xrange(len(secret)):
            if secret[i] != guess[i] and \
                d.get(guess[i]) > 0:
                    cows += 1
                    d[guess[i]] -= 1

        return str(bulls) + 'A' + str(cows) + 'B'
